File: 2020/day03-1.ts
function solution(content: string) {
  return countTrees(content.split("\n"), 0, 0, 3, 1);
}

function countTrees(
  treemap: string[],
  startX: number,
  startY: number,
  deltaX: number,
  deltaY: number
) {
  const width = treemap.at(0)?.length ?? 0;
  const height = treemap.length;
  let currentX = startX;
  let currentY = startY;
  let trees = 0;
  while (currentY < height) {
    if (treemap[currentY][currentX] === "#") trees += 1;
    currentX = (currentX + deltaX) % width;
    currentY += deltaY;
  }
  return trees;
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2020/day03-1.hs
import System.IO

solution :: String -> Int
solution = countTrees (0, 0) (3, 1) . lines

countTrees :: (Int, Int) -> (Int, Int) -> [String] -> Int
countTrees (startX, startY) (deltaX, deltaY) treemap
  | startY < length treemap = if' ((!! startX) ((!! startY) treemap) == '#') 1 0 + countTrees ((startX + deltaX) `mod` length ((!! 0) treemap), startY + deltaY) (deltaX, deltaY) treemap
  | otherwise = 0

if' :: Bool -> a -> a -> a
if' True x _ = x
if' False _ y = y

main :: IO ()
main = do openFile "day03.dat" ReadMode >>= hGetContents >>= print . solution
